modules/tinymce/src/core/main/ts/fmt/FormatUtils.ts

```diff
@@ -169,6 +170,59 @@ const getParents = function (dom: DOMUtils, node: Node, selector?: string) {
   return dom.getParents(node, selector, dom.getRoot());
 };
 
+const isVariableFormatName = (editor: Editor, formatName: string): boolean => {
+  const hasVariableValues = (format: Format) => {
+    const isVariableValue = (val: string): boolean => val.length > 1 && val.charAt(0) === '%';
+    const stylesFieldHas = Obj.get(format, 'styles').exists((stylesField) => Arr.exists(Obj.values(stylesField), isVariableValue));
+    const attributesFieldHas = Obj.get(format, 'attributes').exists((attributesField) => Arr.exists(Obj.values(attributesField), isVariableValue));
+    return stylesFieldHas || attributesFieldHas;
+  };
+  return Arr.exists(editor.formatter.get(formatName) as Format[], (format) => hasVariableValues(format));
+};
+
+/**
+ *  Get all of the format names present on the specifed node
+ */
+const getFormatNamesPresent = (editor: Editor, node: Node, blacklist: string[] = [ 'removeformat' ], deep: boolean = false): string[] => {
+  const validFormats = Arr.filter(Obj.keys(editor.formatter.get()), (name) => !Arr.exists(blacklist, (blacklistedName) => blacklistedName === name));
+  const formats = Arr.map(validFormats, (name) => ({
+    name,
+    matchSimilar: isVariableFormatName(editor, name),
+    checked: false
+  }));
+  const matchedFormatNames: string[] = [];
+
+  const checkNodeForFormats = (currentNode: Node) => {
+    Arr.each(formats, (format) => {
+      if (!format.checked && editor.formatter.matchNode(currentNode, format.name, {}, format.matchSimilar)) {
+        format.checked = true;
```

Comment:
This mutates the format, which is something we generally shouldn't be doing without a good reason. Additionally, a) the `deep` functionality is never used from what I can see, so I'd suggest removing it and b) this logic seems to largely be a duplicate of `matchAll` in `MatchFormat.ts`, so can we reuse that instead?

Reply:
If that logic can't be reused, then we should look at rewriting this to use `Arr.foldl` instead of using `Arr.each` and mutating the data.